Index: lib/backends/io/operations.c
===================================================================
RCS file: /home/cvs/l4/pkg/dietlibc/lib/backends/io/operations.c,v
retrieving revision 1.54
diff -u -r1.54 operations.c
--- lib/backends/io/operations.c	8 Nov 2004 13:27:11 -0000	1.54
+++ lib/backends/io/operations.c	10 Nov 2004 20:23:18 -0000
@@ -689,15 +689,55 @@
     buf->st_ino   = stat_buf.st_ino;
     buf->st_mode  = stat_buf.st_mode;
     buf->st_nlink = stat_buf.st_nlink;
+    buf->st_size  = stat_buf.st_size;
 
     return 0;
 }
 
 int fstat(int fd, struct stat *buf)
 {
-// fixme: do something useful here
-    errno = EBADF;
-    return -1;
+    /* 1. check 'fd'
+     * 2. do a stat on it
+     * 3. convert stat type and return it
+     */
+
+    int ret;
+    l4vfs_stat_t stat_buf;
+    file_desc_t file_desc;
+
+    LOGd(_DEBUG, "fstat(%d, %p)", fd, buf);
+    // 1.
+    if (! ft_is_open(fd))
+    {
+        errno = EBADF;
+        return -1;
+    }
+    LOGd(_DEBUG, "is_open");
+    file_desc = ft_get_entry(fd);
+    if (l4_is_invalid_id(file_desc.server_id))
+    { // should not happen
+        errno = EBADF;
+        return -1;
+    }
+    LOGd(_DEBUG, "is ok");
+
+    // 2.
+    ret = l4vfs_fstat(file_desc.server_id, file_desc.object_handle,&stat_buf);
+    if (ret != 0)
+    {
+        errno = ret;
+        return -1;
+    }
+
+    // 3.
+    // fixme: support more fields here
+    buf->st_dev   = stat_buf.st_dev;
+    buf->st_ino   = stat_buf.st_ino;
+    buf->st_mode  = stat_buf.st_mode;
+    buf->st_nlink = stat_buf.st_nlink;
+    buf->st_size  = stat_buf.st_size;
+
+    return 0;
 }
 
 int lstat(const char* pathname, struct stat *buf)
Index: lib/backends/time/gettimeofday.c
===================================================================
RCS file: /home/cvs/l4/pkg/dietlibc/lib/backends/time/gettimeofday.c,v
retrieving revision 1.5
diff -u -r1.5 gettimeofday.c
--- lib/backends/time/gettimeofday.c	10 Aug 2004 11:32:14 -0000	1.5
+++ lib/backends/time/gettimeofday.c	10 Nov 2004 20:23:21 -0000
@@ -10,6 +10,7 @@
  * GNU General Public License 2. Please see the COPYING file for details.
  */
 #include <sys/time.h>
+#include <time.h>
 
 #include <l4/util/rdtsc.h>
 #include <l4/sys/l4int.h>
@@ -30,5 +31,14 @@
         tv->tv_usec = ns / 1000;
     }
 
+    if (tz)
+    {
+	// man 2 gettimeofday says the use of the 'tz' argument is
+	// deprecated, but we should still fill in something here to
+	// avoid bogus values
+	tz->tz_minuteswest = 0;
+	tz->tz_dsttime = 0;
+    }
+
     return 0;
 }
