Index: lib/backends/io/ioctl.c
===================================================================
RCS file: /home/cvs/l4/pkg/dietlibc/lib/backends/io/ioctl.c,v
retrieving revision 1.7
diff -u -r1.7 ioctl.c
--- lib/backends/io/ioctl.c	10 Aug 2004 11:32:11 -0000	1.7
+++ lib/backends/io/ioctl.c	14 Oct 2004 08:48:53 -0000
@@ -86,6 +86,13 @@
         case SIOCGIFNETMASK:
             count = sizeof (struct ifreq);
             break;
+	case FIONREAD:
+	    // temporary hack: count specifies the size of both the in and
+	    // out arg; this is bad, because for FIONREAD the size of
+	    // the in arg is 0, but it returns an int of size 4
+	    // we need to fix the IDL for cases like this.
+	    count = 4;
+	    break;
         default:
            // by default the data size can be extracted from the command 
             // (thanks to the ioctl macros stolen from linux...:)
